In [9]:
pip install -r requirements.txt

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 914.9/914.9 kB 37.7 MB/s  0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.2/2.2 MB 46.6 MB/s  0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 14.6/14.6 MB 68.7 MB/s  0:00:006m0:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.4/12.4 MB 68.4 MB/s  0:00:00
  Attempting uninstall: jupyterlab━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2/7 [ipywidgets]
    Found existing installation: jupyterlab 4.5.0━━━━━━━━━━━━━ 2/7 [ipywidgets]
    Uninstalling jupyterlab-4.5.0:━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2/7 [ipywidgets]
      Successfully uninstalled jupyterlab-4.5.00m━━━━━━━━━━━━━━━━━ 4/7 [jupyterlab]
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7/7 [jupyter]m5/7 [notebook]b]

[notice] A new release of pip is available: 25.3 -> 26.0.1
[notice] To update, run: pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


In [11]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [12]:
df = pd.read_csv("../Data/batch-1-10.csv")
df.head()

,Time (h),Aeration rate(Fg:L/h),Agitator RPM(RPM:RPM),Sugar feed rate(Fs:L/h),Acid flow rate(Fa:L/h),Base flow rate(Fb:L/h),Heating/cooling water flow rate(Fc:L/h),Heating water flow rate(Fh:L/h),Water for injection/dilution(Fw:L/h),Air head pressure(pressure:bar),...,Oxygen Uptake Rate(OUR:(g min^{-1})),Oxygen in percent in off-gas(O2:O2 (%)),Offline Penicillin concentration(P_offline:P(g L^{-1})),Offline Biomass concentratio(X_offline:X(g L^{-1})),Carbon evolution rate(CER:g/h),Ammonia shots(NH3_shots:kgs),Viscosity(Viscosity_offline:centPoise),Fault reference(Fault_ref:Fault ref),0 - Recipe driven 1 - Operator controlled(Control_ref:Control ref),batch_id
0,0.2,30,100,8,0.0000,30.118,9.8335,0.0001,0,0.6,...,0.480510,0.19595,NaN,NaN,0.034045,0,NaN,0,0,1
1,0.4,30,100,8,0.0000,51.221,18.1550,0.0001,0,0.6,...,0.058147,0.20390,NaN,NaN,0.038702,0,NaN,0,0,1
2,0.6,30,100,8,0.0000,54.302,9.5982,0.0001,0,0.6,...,-0.041505,0.20575,NaN,NaN,0.040240,0,NaN,0,0,1
3,0.8,30,100,8,0.0000,37.816,4.3395,0.0001,0,0.6,...,-0.056737,0.20602,NaN,NaN,0.041149,0,NaN,0,0,1
4,1.0,30,100,8,0.5181,18.908,1.1045,0.0001,0,0.6,...,-0.049975,0.20589,1.017800e-25,0.52808,0.041951,0,4.083,0,0,1


In [ ]:

df.info()
columns = list(df.columns)
print(columns)
df.describe()

<class 'pandas.DataFrame'>
RangeIndex: 11585 entries, 0 to 11584
Data columns (total 34 columns):
 #   Column                                                              Non-Null Count  Dtype  
---  ------                                                              --------------  -----  
 0   Time (h)                                                            11585 non-null  float64
 1   Aeration rate(Fg:L/h)                                               11585 non-null  int64  
 2   Agitator RPM(RPM:RPM)                                               11585 non-null  int64  
 3   Sugar feed rate(Fs:L/h)                                             11585 non-null  int64  
 4   Acid flow rate(Fa:L/h)                                              11585 non-null  float64
 5   Base flow rate(Fb:L/h)                                              11585 non-null  float64
 6   Heating/cooling water flow rate(Fc:L/h)                             11585 non-null  float64
 7   Heating water flow rate(

,Time (h),Aeration rate(Fg:L/h),Agitator RPM(RPM:RPM),Sugar feed rate(Fs:L/h),Acid flow rate(Fa:L/h),Base flow rate(Fb:L/h),Heating/cooling water flow rate(Fc:L/h),Heating water flow rate(Fh:L/h),Water for injection/dilution(Fw:L/h),Air head pressure(pressure:bar),...,Oxygen Uptake Rate(OUR:(g min^{-1})),Oxygen in percent in off-gas(O2:O2 (%)),Offline Penicillin concentration(P_offline:P(g L^{-1})),Offline Biomass concentratio(X_offline:X(g L^{-1})),Carbon evolution rate(CER:g/h),Ammonia shots(NH3_shots:kgs),Viscosity(Viscosity_offline:centPoise),Fault reference(Fault_ref:Fault ref),0 - Recipe driven 1 - Operator controlled(Control_ref:Control ref),batch_id
count,11585.000000,11585.000000,11585.0,11585.000000,11585.000000,11585.000000,11585.000000,11585.000000,11585.000000,11585.000000,...,11585.000000,11585.000000,2.090000e+02,209.000000,11585.000000,11585.0,209.000000,11585.0,11585.0,11585.000000
mean,117.111006,65.250324,100.0,76.611998,0.087966,61.095021,63.610573,21.365101,155.222270,0.944281,...,1.291363,0.191230,1.421598e+01,18.903237,1.254389,0.0,52.224598,0.0,0.0,5.501511
std,68.820049,11.426350,0.0,23.654889,0.595126,48.269333,79.941359,47.498069,155.167953,0.133283,...,0.492681,0.004478,9.868238e+00,7.000869,0.487212,0.0,24.274335,0.0,0.0,2.897838
min,0.200000,30.000000,100.0,8.000000,0.000000,0.000000,0.000100,0.000100,0.000000,0.600000,...,-0.548630,0.156240,1.017800e-25,0.487730,0.032293,0.0,4.080900,0.0,0.0,1.000000
25%,58.000000,60.000000,100.0,76.000000,0.000000,28.834000,12.222000,0.000100,0.000000,0.900000,...,0.962080,0.188280,5.737500e+00,16.928000,0.974250,0.0,35.283000,0.0,0.0,3.000000
50%,116.000000,65.000000,100.0,80.000000,0.000000,55.941000,33.942000,0.042911,100.000000,0.900000,...,1.314100,0.190850,1.477800e+01,21.060000,1.413700,0.0,55.011000,0.0,0.0,6.000000
75%,173.800000,75.000000,100.0,90.000000,0.000000,78.154000,89.764000,12.141000,250.000000,1.100000,...,1.692700,0.193980,2.173800e+01,23.757000,1.641800,0.0,74.894000,0.0,0.0,8.000000
max,278.000000,75.000000,100.0,150.000000,7.798800,225.000000,865.410000,652.170000,500.000000,1.100000,...,5.839300,0.215090,3.450200e+01,26.792000,1.953400,0.0,90.260000,0.0,0.0,10.000000


## Dataset Overview

The dataset contains **11,585 observations and 34 variables** collected from a **batch penicillin fermentation process** in an industrial-style bioreactor. No missing values are present. The variables consist of **23 floating-point measurements and 11 integer variables**.

Each row represents a **time point within a batch**, identified by `batch_id`. The variables describe operating conditions, biological state variables, and measured process outputs.

### Main Variable Categories

**1. Process Operating Conditions (Inputs / Control Variables)**
These describe how the reactor is operated and will later serve as key optimization variables.

* Aeration rate (L/h)
* Agitator RPM
* Sugar feed rate (L/h)
* Acid flow rate (L/h)
* Base flow rate (L/h)
* Heating/cooling water flow rate (L/h)
* Heating water flow rate (L/h)
* Water for injection/dilution (L/h)
* Air head pressure (bar)

**2. Bioprocess State Variables**
These describe the internal biological and chemical state of the fermentation.

* Substrate concentration (g/L)
* Dissolved oxygen concentration (mg/L)
* pH
* Temperature (K)
* Vessel volume (L)
* Vessel weight (kg)
* Viscosity (centipoise)

**3. Production and Biological Performance Variables**
These variables describe microbial activity and product formation.

* Penicillin concentration (g/L)
* Biomass concentration (offline, g/L)
* Oxygen uptake rate (g/min)
* Carbon evolution rate (g/h)
* CO₂ percentage in off-gas (%)
* O₂ percentage in off-gas (%)

**4. Feed and Chemical Additions**

* PAA flow (L/h)
* PAA concentration (g/L)
* Oil flow (L/h)
* Ammonia concentration (g/L)
* Ammonia shots (kg)

**5. Operational / Monitoring Variables**

* Generated heat (kJ)
* Dumped broth flow (L/h)
* Fault reference
* Control reference (recipe driven vs operator controlled)

### Units

Most variables use **standard bioprocess engineering units**, including:

* **Flow rates:** L/h
* **Concentrations:** g/L or mg/L
* **Pressure:** bar
* **Temperature:** Kelvin (K)
* **Gas composition:** %
* **Biological rates:** g/min or g/h
* **Viscosity:** centipoise

### Relevance to the Project

This dataset provides both **process control variables (RPM, aeration, feed rates)** and **biological response variables (biomass, penicillin concentration)**.

These features will be used to:

* **Model fermentation dynamics** over time
* **Predict biomass and penicillin concentration trajectories**
* **Identify optimal operating conditions** that maximize final penicillin yield
* **Simulate alternative operating strategies** for process optimization

The presence of detailed operational variables and gas-exchange measurements (OUR, CER, off-gas composition) makes the dataset suitable for **feature engineering, dynamic modeling, and future integration with mechanistic bioprocess models or digital twins**.
